### PR TITLE
Documentation of UD_CustomDevice_RenderScript

### PR DESCRIPTION
**Pull request by IHateMyKite**
- Add documentation of UD_CustomDevice_RenderScript
- Clean up mess along the way

### CODE REVIEW

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 1]
```diff
@@ -136,12 +139,64 @@ EndFunction
...
+;/  Variable: DeviceRendered
+    Link to device the script is currently on. In case this is not filled, it will be taken from Inventory Device
+/;
+Armor    Property DeviceRendered                                auto hidden ;Is taken from ID
+
+
+String[] Property UD_DeviceStruggleKeywords                     auto Hidden ;keywords (as string array) used to filter struggle animations
+
+;/  Variable: UD_ActiveEffectName
+    Name of active effect. Defaultly set to "Share"
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:182]: Name of active effect. By default set to "Share"

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 2]
```diff
@@ -136,12 +139,64 @@ EndFunction
...
+
+;/  Variable: UD_DeviceType
+    Name of the device type. Only used in details, so user can distinguish various device types
+/;
+string   Property UD_DeviceType                 = "Generic"     auto hidden ;name of the device type
+
+;/  Variable: UD_DeviceKeyword_Minor
+    Minor device keyword. Only used by animations
+    
+    In most cases, there is not need to chenge this, as the framework will take the keyword automatically based on main keyword
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:198]: ... there is no need to change this...

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 3]
```diff
@@ -171,7 +226,29 @@ Keyword     Property UD_DeviceKeyword_Minor                 ;minor keyword of th
...
 
+;/  Group: Customization
+===========================================================================================
+===========================================================================================
+===========================================================================================
+/;
+
 ;MAIN VALUES
+;/  Variable: UD_Level
+    Level of device. Defaultly 1.
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:237]: Level of device. By default set to 1.

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 4]
```diff
@@ -193,6 +286,19 @@ float       Property UD_durability_damage_base                              ;dur
         return decodeBit(_deviceControlBitMap_7,12,0)/100.0
     EndFunction
 EndProperty
+
+;/  Variable: UD_base_stat_drain
+    How many points of stats (helth, stamina, magicka) are reduced per second of minigame. This is only base values, which is later moded with minigame values
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:291]: ...health...

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 1 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:182]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 2 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:198]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 3 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:237]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 4 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:291]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 5]
```diff
@@ -2829,188 +3123,552 @@ EndFunction
...
-float Function getDurabilityDmgMod()
-    return _durability_damage_mod
-EndFunction
-
-;/  Function: updateDifficulty
-    Update device difficulty. Updates return value of <getDurabilityDmgMod>. Validate jammed locks
+;/  About: Control Array
+    Control array is used for filtering menu options. The value is inverted. So if you want to disable one of the options, you have to change nth value to True
+    
+    Defaultly all array values are False, so if you create the array, you can only disable options that you don't want, while ignoring other options
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:3136]: All array values set to False by default, so...

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 6]
```diff
@@ -2829,188 +3123,552 @@ EndFunction
...
+    
+    *Allways needs to have size 18!!*
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:3138]: Always

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 7]
```diff
@@ -3245,672 +3903,216 @@ float Function getMaxDurability()
...
+        Get absolute condition of device. Always in range from 0.0 to max device health
+/;
 float Function getCondition()
     return _total_durability_drain
 EndFunction
 
+;/  Function: getRelativeCondition
+    Returns:
+
+        Get relative condition of device. Always in range from 0.0 to max device health
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:3918]: Isn't it returning value in range of 0.0..1.0 here?

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 8]
```diff
@@ -3245,672 +3903,216 @@ float Function getMaxDurability()
...
-        elseif msgChoice == 6    ;command
-            aControl = new Bool[30]
-            DeviceMenu(aControl)
-            _break = True
-        elseif msgChoice == 7    ;details
-            processDetails()        
-        else
-            _break = True        ;exit
-        endif
+        afBase      - Bse of resistence (for example 1.0=100%)
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:4093]: Base of resistance

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 9]
```diff
@@ -3245,672 +3903,216 @@ float Function getMaxDurability()
...
-        int  loc_res  = UD_SpecialMenuInteractionWH.show()
-        bool loc_res2 = proccesSpecialMenuWH(akSource,loc_res)
-        return loc_res2
-    else
-        return False
-    endif
+;/  Function: getModResistMagicka
+    Parameters:
+
+        afBase      - Bse of resistence (for example 1.0=100%)
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:4107]: Base of resistance

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 10]
```diff
@@ -3245,672 +3903,216 @@ float Function getMaxDurability()
...
+        afCondMod   - Condition modifier. Change value by (10% + afCondMod) per condition level
+        
+    Returns:
+
+        Magick damage multiplier
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:4112]: Magicka or Magic

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 5 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:3136]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 6 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:3138]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 7 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:3918]: You are right. I copied it from getCondition and forget to change it. Fixed.

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 8 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:4093]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 9 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:4107]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 10 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:4112]: FIxed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 11]
```diff
@@ -2078,327 +2078,585 @@ bool Function helperFreeLegs()
...
+        return getModifierAllParam(asModifier).length
+    endif
+EndFunction
+
+;/  Function: getModifierIntParam
+    Parameters:
+
+        asModifier      - Name of modifier
+        aiIndex         - Index of parameter
+        aiDefaultValue  - Defualt value in case of error
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:2533]: Default

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 12]
```diff
@@ -2078,327 +2078,585 @@ bool Function helperFreeLegs()
...
+        return loc_params[aiIndex] as Int
+    endif
+EndFunction
+
+;/  Function: getModifierFloatParam
+    Parameters:
+
+        asModifier      - Name of modifier
+        aiIndex         - Index of parameter
+        afDefaultValue  - Defualt value in case of error
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:2557]: Default

File: Scripts/Source/UD_CustomDevice_RenderScript.psc [hunk 13]
```diff
@@ -2078,327 +2078,585 @@ bool Function helperFreeLegs()
...
+        return loc_params[aiIndex] as Float
+    endif
+EndFunction
+
+;/  Function: getModifierParam
+    Parameters:
+
+        asModifier      - Name of modifier
+        aiIndex         - Index of parameter
+        afDefaultValue  - Defualt value in case of error
```
Comment (kurotatsu77) [inline Scripts/Source/UD_CustomDevice_RenderScript.psc:2581]: Default

File: docs/RenderDeviceTemplate.psc [hunk 14]
```diff
@@ -15,163 +48,965 @@ EndFunction
...
+    Editing this function allows to create custom effect when device is lockpicked
+    
+    Should always call parent script, to prevent unintended behaviour
+/;
 Function OnDeviceLockpicked() ;called when device is lockpicked
     parent.OnDeviceLockpicked()
 EndFunction
+
+;/  Function: OnLockReached
+    *!!NO LOGNER USED!!*
```
Comment (kurotatsu77) [inline docs/RenderDeviceTemplate.psc:489]: NO LONGER

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 11 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:2533]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 12 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:2557]: Fixed

File: Scripts/Source/UD_CustomDevice_RenderScript.psc (hunk 13 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_CustomDevice_RenderScript.psc:2581]: Fixed

File: docs/RenderDeviceTemplate.psc (hunk 14 above)
Comment (IHateMyKite) [reply docs/RenderDeviceTemplate.psc:489]: Fixed

**Review by kurotatsu77 — APPROVED** [review]
Massive work, I hope I didn't missed a lot of errors.

**Pull request merged**